### PR TITLE
K8stopology

### PR DESCRIPTION
**Pull request by craffit**

### CODE REVIEW

File: checks.d/kubernetes_topology.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+# 3rd party
+
+# project
+from checks import AgentCheck
+from utils.kubernetes import KubeUtil
+
+class KubernetesTopology(AgentCheck):
+    INSTANCE_TYPE = "kubernetes"
+    SERVICE_CHECK_NAME = "kubernetes.topology_information"
+    service_check_needed = True
```
Comment (whazenberg) [inline checks.d/kubernetes_topology.py]: unused.

File: checks.d/kubernetes_topology.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+# 3rd party
+
+# project
+from checks import AgentCheck
+from utils.kubernetes import KubeUtil
+
+class KubernetesTopology(AgentCheck):
+    INSTANCE_TYPE = "kubernetes"
+    SERVICE_CHECK_NAME = "kubernetes.topology_information"
```
Comment (whazenberg) [inline checks.d/kubernetes_topology.py:19]: unused.

File: checks.d/kubernetes_topology.py (hunk 1 above)
Comment (whazenberg) [reply checks.d/kubernetes_topology.py]: Either we can add a self.service_check(..) or remove the unused code.

**Review by whazenberg — CHANGES_REQUESTED** [review]

**Comment by whazenberg** [comment]
Incomplete review

File: checks.d/kubernetes_topology.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        if not self.kubeutil.host:
+            raise Exception('Unable to retrieve Docker hostname and host parameter is not set')
+
+    def check(self, instance):
+        # For now we only support monitoring 1 kubernetes at a time, so the instance_key containing
+        # just the 'kubernetes' type is sufficient. Once we start supporting monitoring multiple clusters
+        # simultaneously we need to uniquely identify them in the instance_key
+        instance_key = {
+            'type': self.INSTANCE_TYPE,
+            'url': 'http://kubernetes'
```
Comment (whazenberg) [inline checks.d/kubernetes_topology.py]: we should use the url of the instance here, which should be set in the config

File: utils/kubernetes/kubeutil.py [hunk 4]
```diff
@@ -136,6 +155,18 @@ def retrieve_metrics(self):
         """
         return retrieve_json(self.metrics_url)
 
+    def retrieve_nodes_list(self):
+        """
+        Retrieve the list of nodes for this cluster qurying the kublet API.
```
Comment (whazenberg) [inline utils/kubernetes/kubeutil.py]: typo in qurying

File: utils/kubernetes/kubeutil.py [hunk 5]
```diff
@@ -136,6 +155,18 @@ def retrieve_metrics(self):
...
+        """
+        return retrieve_json(self.nodes_list_url)
+
+    def retrieve_services_list(self):
+        """
+        Retrieve the list of services for this cluster qurying the kublet API.
```
Comment (whazenberg) [inline utils/kubernetes/kubeutil.py]: typo in qurying

File: .travis.yml [hunk 6]
```diff
@@ -4,6 +4,7 @@ branches:
   only:
     - master
     - serviceDiscoveryTest
+    - k8stopology
```
Comment (whazenberg) [inline .travis.yml]: let's remove this before merging to master

File: conf.d/kubernetes_topology.yaml.example [hunk 7]
```diff
@@ -0,0 +1,15 @@
+init_config:
+  #    tags:
+  #      - optional_tag1
+  #      - optional_tag2
+
+instances:
+  # The kubernetes topology check retrieves metrics from the kubernetes api.
```
Comment (whazenberg) [inline conf.d/kubernetes_topology.yaml.example]: retrieves metrics?

**Review by whazenberg — CHANGES_REQUESTED** [review]

File: checks.d/kubernetes_topology.py (hunk 3 above)
Comment (raboof) [reply checks.d/kubernetes_topology.py]: it should be unique to the cluster, not just to the instance, right?

File: checks.d/kubernetes_topology.py [hunk 8]
```diff
@@ -119,3 +123,7 @@ def _link_pods_to_services(self, kubeutil, instance_key):
                         data = dict()
                         pod_name = address['targetRef']['name']
                         self.relation(instance_key, pod_name, service_name, {'name': 'BELONGS_TO'}, data)
+
+    def _flatten_dict(self, dict_of_list):
```
Comment (whazenberg) [inline checks.d/kubernetes_topology.py]: maybe move this to utils?

File: tests/checks/mock/test_kubernetes_topology.py [hunk 9]
```diff
@@ -31,8 +31,8 @@ def test_kube_topo(self, *args):
             'type': 'kubernetes',
             'url': 'http://kubernetes'
         })
-
-        self.assertEqual(len(instances[0]['relations']), 66)
+        # print instances[0]['relations']
```
Comment (whazenberg) [inline tests/checks/mock/test_kubernetes_topology.py]: remove?

**Review by whazenberg — APPROVED** [review]

File: checks.d/kubernetes_topology.py (hunk 2 above)
Comment (whazenberg) [reply checks.d/kubernetes_topology.py:19]: used now

File: checks.d/kubernetes_topology.py (hunk 1 above)
Comment (whazenberg) [reply checks.d/kubernetes_topology.py]: used now

**Pull request merged**